pkg/controller/master/virtualmachine/vmi_controller.go

```diff
@@ -0,0 +1,141 @@
+//   but they are no longer used by the newly created VirtualMachineInstance.
+//   Since the KubeVirt doesn't handle this relationship currently, it may cause a logic leak.
+//
+//   ```diff
+//     apiVersion: kubevirt.io/v1alpha3
+//     kind: VirtualMachine
+//     spec:
+//   +   # the data-disk-2 DataVolume still be referred to this VirtualMachine until delete the whole VirtualMachine.
+//   -   dataVolumeTemplates:
+//   -   - metadata:
+//   -     name: data-disk-2
+//   -     spec:
+//   -       pvc:
+//   -         resource:
+//   -           requests:
+//   -             storage: 10Gi
+//   -       source:
+//   -         blank: {}
+//       template:
+//         spec:
+//           domain:
+//             devices:
+//             - disk:
+//               name: root-disk
+//             - disk:
+//               name: data-disk-1
+//   -         - disk:
+//   -           name: data-disk-2
+//           volumes:
+//           - dataVolume:
+//             name: root-disk
+//           - dataVolume:
+//             name: data-disk-1
+//   -       - dataVolume:
+//   -         name: data-disk-2
+//   ```
+func (h *VMIController) UnrefUselessDataVolumes(_ string, vmi *kubevirtapis.VirtualMachineInstance) (*kubevirtapis.VirtualMachineInstance, error) {
+	if vmi == nil || vmi.DeletionTimestamp == nil {
+		return vmi, nil
+	}
+
+	// NB(thxCode): if the VirtualMachineInstance is generated by VirtualMachine,
+	// it must have an OwnerReference ref to a VirtualMachine,
+	// so we can validate this reference and process this kind of VirtualMachineInstance.
+	var vmReferred = ref.GetControllerOfNoCopy(vmi)
+	if vmReferred == nil {
+		// doesn't process headless VirtualMachineInstance
+		return vmi, nil
+	}
+	var vmGVK = kubevirtapis.VirtualMachineGroupVersionKind
+	if vmReferred.APIVersion != vmGVK.GroupVersion().String() ||
+		vmReferred.Kind != vmGVK.Kind {
+		// doesn't process none VirtualMachine type managed VirtualMachineInstance
+		return vmi, nil
+	}
+
+	var vm, err = h.virtualMachineCache.Get(vmi.Namespace, vmReferred.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// NB(thxCode): in this case, the related VirtualMachine is deleted,
+			// which triggers the cascading deletion of the VirtualMachineInstance.
+			// however, the deletion of the VirtualMachineInstance is placed in the background,
+			// the VirtualMachineInstance is still deleting, but the corresponding VirtualMachine has been deleted.
+			// so we can ignores this error.
+			return vmi, nil
+		}
+		return vmi, fmt.Errorf("failed to get VirtualMachine referred by VirtualMachineInstance(%s/%s): %w", vmi.Namespace, vmi.Name, err)
+	}
+	if vm.DeletionTimestamp != nil {
+		// doesn't process as there is another unref reconciling executed by VirtualMachine controller
+		return vmi, nil
+	}
+
+	// collects all desired DataVolumes from VirtualMachine
+	var dataVolumeNamesDesired = sets.String{}
```

Comment:
prefer to call `dataVolumeNames` as the comment already explains it.

Reply:
Updated